### ISSUE TITLE
[Question] Difference between "architecture style"?

### ISSUE THREAD

**Issue by johannesschobel** [nstudio/xplat#189]
Dear @NathanWalker ,

thank you for this wonderful add-on to `nrwl/nx` to allow using `ìonic` within the `nrwl` scope! Further, i would like to thank you for your time invested in migrating `xplat` to the latest `angular` version.

In the issue dealing with the `ionic 5 angular 9 migration`, we discussed that a lot of problems arise from using the `xplat supporting architecture` ( https://github.com/nstudio/xplat/issues/180 ).

What exactly is the `xplat supporting architecture` in this context? And what is the difference to the "regular architecture"?

-----

### Regular Architecture 
- Just a plain ionic 5 application with everything in the `/apps` folder.

### xplat Architecture
- There is - more or less - an "empty" `/apps` project that includes stuff from the `/libs` folder. 
- Further, the `/libs` folder now contains folders for `ui`, `feat`, and so on..

-----

Are there any other differences than just the "folder structure"? Because i would like to use my own structure in the `libs` folder. My `nrwl/nx` workspace already contains a `libs` folder with a lot of different sub-folders and modules. Therefore, i cannot use the `xplat supporting architecture`, right?

All the best and thank you very much for your time answering this question!

**Comment by phillipplum** [comment]
Hello @johannesschobel,
I have now played around with the xplat architecture.
When I not using the xplat architecture, in the app folder the `tsconfig.json` only has
```
"paths": {
   "@angular/*": ["../../node_modules/@angular/*"]
}
```

When I use the xplat architecture in the app folder `tsconfig.json` now you can find:
```
"paths": {
      "@myworkspace/*": ["../../libs/*"],
      "@myworkspace/ionic": ["../../xplat/ionic/index.ts"],
      "@myworkspace/ionic/*": ["../../xplat/ionic/*"],
      "@myworkspace/web": ["../../xplat/web/index.ts"],
      "@myworkspace/web/*": ["../../xplat/web/*"],
      "@angular/*": ["../../node_modules/@angular/*"]
    }
```

This creates the problem for me that after each creation of a new app, I have to manually add the missing lines in the `tsconfig.json`.

**Comment by NathanWalker** [comment]
@phillipplum why would you need to add the lines manually in tsconfig.json? 
Generating with or without xplat architecture sets up your tsconfig in the app correctly as mentioned above?

**Comment by NathanWalker** [comment]
@johannesschobel always a great question.
This explains the reason for it fairly well:
https://nstudio.io/xplat/fundamentals/architecture

But let me expand on a few things with a few questions:

* If you build and deploy an Ionic app and in the future your company wants to deploy another Ionic app internally for different purpose but want it to generally have same vibe/style. You want to reuse a lot of the styles (css or scss), what is your plan on reusing a lot of scss/css work you did for that app and use as a baseline in another Ionic app which would speed up development by just allowing you to theme the other app on top of what you already did? Would the plan be copy/paste and/or duplication of styling work into the other app in the workspace? And beyond that what would your plan be on keeping maintenance costs down in maintaining those 2 apps now vs. just the one?

* Let's say in a year your company decides it wants and maybe needs to deploy a desktop Electron app. You and your team invested a lot of time in building your Ionic app and would prefer to not retrofit everything to deal with Electron specifics. What would your plan be to deal with such a request efficiently and without pain? Given the common case focus was solely on just Ionic development and not thinking about future needs which ended up coupling a lot of Ionic behavior into low level core services in your workspace? 

* Consider the above but this time it's iOS/Android native with something like NativeScript? How could you help future proof some of the workspace today to help deal with the unforeseen?

This is largely what the `xplat` layer is about in addition to the baseline set added to libs. It's intended to be built on top of and around but attempts to imply a consistent approach to various cross platform needs (needs that are often not known today but could come up in future).
By preparing the codebase with common denominator services like `WindowService` and a baseline `BaseComponent` which bakes in a consistent way to handle subscription teardown's throughout anything built in the codebase it helps prepare from the beginning several setups that can ease some of that pain. For example `WindowService` can help bake in an approach to handling global `window` in ways that often cause quite a bit of pain when comes to adding server side rendering at a later time and/or even NativeScript.

Given the new internal Nx plugin structure that was introduced for plugins (mentioned here: https://nx.dev/angular/guides/nx-plugin) we've been looking into whether that structure would solve the needs that we have found for our own client projects. There's possibility in future we could explore publishing some of the tools for that structure as an alternate route.

Hope that makes sense.

**Comment by phillipplum** [comment]
> @phillipplum why would you need to add the lines manually in tsconfig.json?
> Generating with or without xplat architecture sets up your tsconfig in the app correctly as mentioned above?

Maybe that shows the problem or my wrong way of doing it.

I create a workspace and create a lib: `test`. Now I'm creating an ionic-angular app: `dummy`. I do all of this without the xplat architecture.
At this point my root `tsconfig.json` has the following path settings:
```
 "paths": {
   "@myworkspace/test": ["./libs/test/src/index.ts"]
}
```
and my `tsconfig.json` in `apps/ionic-dummy` following content:
```
"paths": {
   "@angular/*": ["../../node_modules/@angular/*"]
}
```

And now I have the problem, that I can not import from @myworkspace/test in my ionic-angular app `dummy`.

**Comment by NathanWalker** [comment]
Thanks @phillipplum that helps explain - When creating app's without xplat architecture and plan to use purely Nx lib's yes you are going a more manual route and do have to configure your tsconfig appropriately for those cases. We could look into automatically updating the Ionic app's tsconfig in case of pure Nx lib creation. Sounds like that could help.

**Comment by johannesschobel** [comment]
Dear @NathanWalker , thank you for your detailed information on the `xplat supporting architecture`. Moving parts of the application to the `libs` folder is a known "pattern" from `nrwl/nx` - so i am quite familiar with the latter.

However, what is the difference between the `libs/` folder and the `xplat/*` folders?

Is the "thought process" of `xplat` to put "features" **only relevant to xplat specific applications** (i.e., electron, ionic, ...) in the `xplat/ionic/core` folder? Whereas `basic core` functionality (that may be used across the overall platform (web, api, mobile) into the `libs/core` folder?

Maybe you can put some more light on the difference between `libs/` and `xplat/*/` folders?

Thank you very much and stay safe

**Comment by NathanWalker** [comment]
@johannesschobel That's correct. `libs` is intended to be "platform agnostic" (meaning TypeScript code that can be executed on *any* platform (no specific platform details within) and `xplat` is intended to be "platform dependent" keeping a clean separation of platform boundaries making the code portable and more flexible as time moves on and needs as well as industry changes.

We have long wanted to add some extra Nx module enforcements to these rules so the editor would flag code to the developer if someone accidentally brokes those rules to help keep platform separation tidy as well as easier to maintain.
